### DISCUSSION TITLE
Couple questions

### DISCUSSION THREAD

**Discussion by gbrits** [capacitor-community/google-maps#32] [Q&A]
Hi guys, firstly - thanks a lot for building this out, it has so far met the requirements for my app!

Just a few questions:

Can I fitBounds to my markers?

It possible to style the map?

Is it possible to draw polylines between markers? 

Would it be possible to create a dragging left to right slider that would then show up markers & draw a line between them?

Any assistance would be appreciated!

**Comment by gbrits** [comment]
Managed to set a polyline along all my points, with:

```
var points = [];
for (var i = 0; i < markers.length; i++) {
    points.push({
        longitude: Number(markers[i].longitude),
        latitude: Number(markers[i].latitude)
    });
}

CapacitorGoogleMaps.addPolyline({ points: points });
```

Doesn't appear to be something I can style, though. (Stroke width etc.)

**Reply by gbrits** [reply]
Getting this error, when attempting to style the map:

```
[error] - Unhandled Promise rejection: undefined is not a function (near '...CapacitorGoogleMaps.setMapStyle...') ; Zone: <root> ; Task: null ; Value: {}
```

Both with:
```
CapacitorGoogleMaps.setMapStyle("[{\"featureType\":\"all\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"}]},{\"featureType\":\"administrative\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"}]},{\"featureType\":\"landscape\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"},{\"hue\":\"#ff0000\"}]},{\"featureType\":\"poi.business\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"},{\"lightness\":\"60\"},{\"saturation\":\"7\"},{\"weight\":\"0.20\"}]}]")
```
As well as
```
CapacitorGoogleMaps.setMapStyle({ jsonString: "[{\"featureType\":\"all\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"}]},{\"featureType\":\"administrative\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"}]},{\"featureType\":\"landscape\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"},{\"hue\":\"#ff0000\"}]},{\"featureType\":\"poi.business\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"},{\"lightness\":\"60\"},{\"saturation\":\"7\"},{\"weight\":\"0.20\"}]}]" })
```

**Reply by gbrits** [reply]
With a combination of using CapacitorGoogleMaps.clear() and CapacitorGoogleMaps.setMarker I was able to make a percentage based slider show & hide markers dependant on the "progress" dragEnd. (using gsap draggable library)

```
Draggable.create(".dragger", {
              type: "x",
              bounds: ".draggerBounds",
              onDrag: function() {
                let percent = Number((this.x / this.maxX) * 100).toFixed(0);
                let sliced_points = points.slice(0, (points.length * percent / 100).toFixed(0));
                let sliced_markers = that.entry.gps_points.slice(0, (that.entry.gps_points.length * percent / 100).toFixed(0));
                CapacitorGoogleMaps.clear();
                CapacitorGoogleMaps.addPolyline({ points: sliced_points });
                sliced_markers.forEach(iterate);
              }
            });
```

**Reply by hemangsk** [reply]
> Getting this error, when attempting to style the map:
> 
> ```
> [error] - Unhandled Promise rejection: undefined is not a function (near '...CapacitorGoogleMaps.setMapStyle...') ; Zone: <root> ; Task: null ; Value: {}
> ```
> 
> Both with:
> 
> ```
> CapacitorGoogleMaps.setMapStyle("[{\"featureType\":\"all\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"}]},{\"featureType\":\"administrative\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"}]},{\"featureType\":\"landscape\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"},{\"hue\":\"#ff0000\"}]},{\"featureType\":\"poi.business\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"},{\"lightness\":\"60\"},{\"saturation\":\"7\"},{\"weight\":\"0.20\"}]}]")
> ```
> 
> As well as
> 
> ```
> CapacitorGoogleMaps.setMapStyle({ jsonString: "[{\"featureType\":\"all\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"}]},{\"featureType\":\"administrative\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"}]},{\"featureType\":\"landscape\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"on\"},{\"hue\":\"#ff0000\"}]},{\"featureType\":\"poi.business\",\"elementType\":\"all\",\"stylers\":[{\"visibility\":\"simplified\"},{\"lightness\":\"60\"},{\"saturation\":\"7\"},{\"weight\":\"0.20\"}]}]" })
> ```


Can you try using JSON.stringify? This works for me:

```
CapacitorGoogleMaps.setMapStyle({
      jsonString: JSON.stringify([
        {
          "featureType": "all",
          "stylers": [
            { "color": "#C0C0C0" }
          ]
        }, {
          "featureType": "road.arterial",
          "elementType": "geometry",
          "stylers": [
            { "color": "#CCFFFF" }
          ]
        }, {
          "featureType": "landscape",
          "elementType": "labels",
          "stylers": [
            { "visibility": "off" }
          ]
        }
      ])
    })
```

**Reply by hemangsk** [reply]
Just noticed setMapStyle only working with android, needs an iOS fix. Thanks for reporting this!

**Reply by hemangsk** [reply]
Fix released in 0.0.5

**Reply by gbrits** [reply]
Awesome, guess it pays to interact with repositories in discussion :) I'd love to contribute, but I have absolutely no idea how to edit the swift side of things where you control the native side of things. Any resources you can point me to that would be helpful!

Also, a couple of other things I'm not getting right at the moment -- fitting around the boundary of markers, fitBounds or something similar - is there a way to achieve that with this library? Also is there any way to invoke the snippet info of a marker to show programatically? Also, how to change the zoom level? It's not working for me on iOS

**Reply by hemangsk** [reply]
No worries, I'm also a novice with swift side of things. https://capacitorjs.com/docs/plugins/ios is a holy grail and good first guide. It goes over the concept of "bridge". The guide covers all aspects of adding native functionality. so for example,  in order to add / debug something like setMapStyle, this is the list of files we need to deal with:

- https://github.com/hemangsk/capacitor-googlemaps-native/blob/master/src/definitions.ts#L111
- https://github.com/hemangsk/capacitor-googlemaps-native/blob/master/ios/Plugin/Plugin.m#L11
- https://github.com/hemangsk/capacitor-googlemaps-native/blob/master/ios/Plugin/Plugin.swift#L314

Re other questions:
- I think the functionality of fit bounds does not exist yet in this plugin, but not impossible to add.
- Checkout setCamera for zoom level changing

**Comment by gbrits** [comment]
These are great resources, I’m reading the Swift side now. One last thing.. can you explain how you edit an existing project like yours locally? Like if I wanted to make adjustments to this repository, test it locally with the swift changes & then commit the fixes?

**Reply by hemangsk** [reply]
I'm following the approach at https://capacitorjs.com/docs/plugins/workflow#local-testing for local development. To summarise:

- Create an ionic-capacitor project and install the local copy of the plugin using npm link. npm link ensures I install the local version of this library from my local machine's file path, and not from npm.
- Fire up this project, preferably with live reload 
``` ionic capacitor run ios --livereload```
will open project in XCode, we can run this in an emulator.

- Next, we can open capacitor-googlemaps-native/ios/Plugin.**xcworkspace** in Xcode. Now any changes done in the Plugin.swift file will also show up in our dummy project's node_modules/capacitor-googlemaps-native, which means we can instantly see the changes on the emulator / device after doing command R on that ionic project's XCode window. 

Hope that makes sense? Let me know if you've any questions

**Reply by gbrits** [reply]
If you're still around @hemangsk - I'm still struggling just with the setCamera / zooming to a specific part of the map, any idea what I'm doing wrong, or is there maybe an iOS glitch with this?

```
var points = [];
for (var i = 0; i < markers.length; i++) {
    points.push({
        longitude: Number(markers[i].longitude),
        latitude: Number(markers[i].latitude)
    });
}
CapacitorGoogleMaps.setCamera({
    latitude: points[0].latitude,
    longitude: points[0].longitude,
    zoom: 7,
    animate: true
});
```

**Reply by hemangsk** [reply]
Looks like a bug within the package. I'll have to look into it. Incase you want to debug it, Plugin.swift https://github.com/hemangsk/capacitor-googlemaps-native/blob/master/ios/Plugin/Plugin.swift#L287 is where we need to look

**Reply by gbrits** [reply]
I think it's supposed to be longitude & latitude nested inside of a "target" object?
target.longitude
target.latitude

**Reply by gbrits** [reply]
➜  gbrits git:(master) ✗ `npm link capacitor-googlemaps-native@latest`
+ capacitor-googlemaps-native@0.0.5
added 3 packages from 3 contributors in 2.337s

Is this now my local copy?

**Reply by hemangsk** [reply]
To ensure if npm link is setup correctly, check if the package.json of your ionic project has the "file" prefix in the version field of "capacitor-googlemaps-native",  
```
"dependencies": {
...
    **"capacitor-googlemaps-native": "file:../capacitor-googlemaps-native",**
...
  }```

**Reply by hemangsk** [reply]
Ok I think I understand the problems with setCamera, its actually not declared properly. The definitions are missing from definition.ts and ios/Plugin.m

**Reply by gbrits** [reply]
I was way off then, and after using npm link it doesn't show file:... it still shows 0.0.5. Am I supposed to git clone or anything of the sort before editing the library? or can I just edit the files (via xcode) under Pods > Development Pods. Or am I doing it totally wrong? Sorry for the nagging questions.

**Reply by hemangsk** [reply]
Yeah, you need to have a local copy of the plugin via git clone. Then just follow the steps at https://capacitorjs.com/docs/plugins/workflow#local-testing 

Yes we can edit the files directly under Development pods, but ofc if you don't do npm link with a local copy of the plugin before editing under development pods, all the edits will just remain inside your node_modules/capacitor-googlemaps-native and won't be "commit-able". In order to make sure you can push your edits to git later, the local copy is needed.

Ah no worries about the questions, We just unraveled 2 bugs during this discussion so I'll call it a productive day :D
